File: ios/Classes/MetaWearablesPlugin.swift
import Flutter
import Foundation

#if canImport(MWDATCore)
import MWDATCore
import MWDATCamera
#endif

public class MetaWearablesPlugin: NSObject, FlutterPlugin, UIApplicationDelegate {
  private var api: MetaWearablesHostApiImpl?

  public static func register(with registrar: FlutterPluginRegistrar) {
    let messenger = registrar.messenger()
    let api = MetaWearablesHostApiImpl(binaryMessenger: messenger)
    WearablesHostApiSetup.setUp(binaryMessenger: messenger, api: api)
    let instance = MetaWearablesPlugin()
    instance.api = api
    registrar.addApplicationDelegate(instance)
  }

  public func application(
    _ application: UIApplication,
    open url: URL,
    options: [UIApplication.OpenURLOptionsKey: Any] = [:]
  ) -> Bool {
    #if canImport(MWDATCore)
    if URLComponents(url: url, resolvingAgainstBaseURL: false)?
      .queryItems?
      .contains(where: { $0.name == "metaWearablesAction" }) == true {
      Task { _ = try? await Wearables.shared.handleUrl(url) }
      return true
    }
    #endif
    return false
  }
}

private class MetaWearablesHostApiImpl: NSObject, WearablesHostApi {
  private let flutterApi: WearablesFlutterApi
  #if canImport(MWDATCore)
  private var registrationTask: Task<Void, Never>?
  private var streamSession: StreamSession?
  private var streamTasks: [Task<Void, Never>] = []
  private var listenerTokens: [AnyListenerToken] = []
  #endif

  init(binaryMessenger: FlutterBinaryMessenger) {
    flutterApi = WearablesFlutterApi(binaryMessenger: binaryMessenger)
  }

  deinit {
    #if canImport(MWDATCore)
    registrationTask?.cancel()
    stopStream()
    #endif
  }

  func initialize() throws {
    #if canImport(MWDATCore)
    try Wearables.configure()
    startRegistrationStream()
    #endif
  }

  func checkPermission(permission: Permission, completion: @escaping (Result<PermissionResult, Error>) -> Void) {
    #if canImport(MWDATCore)
    Task {
      do {
        let status = try await Wearables.shared.checkPermissionStatus(.camera)
        completion(.success(status.toPigeonResult()))
      } catch {
        completion(.failure(error))
        flutterApi.onError(errorInfo: ErrorInfo(code: "permission", message: error.localizedDescription))
      }
    }
    #else
    completion(.success(PermissionResult(status: .error, message: "MWDATCore not linked. Add the Meta Wearables Swift Package to your app target.")))
    #endif
  }

  func requestPermission(permission: Permission, completion: @escaping (Result<PermissionResult, Error>) -> Void) {
    #if canImport(MWDATCore)
    Task {
      do {
        let status = try await Wearables.shared.requestPermission(.camera)
        completion(.success(status.toPigeonResult()))
      } catch {
        completion(.failure(error))
        flutterApi.onError(errorInfo: ErrorInfo(code: "permission", message: error.localizedDescription))
      }
    }
    #else
    completion(.success(PermissionResult(status: .error, message: "MWDATCore not linked. Add the Meta Wearables Swift Package to your app target.")))
    #endif
  }

  func getRegistrationState(completion: @escaping (Result<RegistrationUpdate, Error>) -> Void) {
    #if canImport(MWDATCore)
    completion(.success(Wearables.shared.registrationState.toPigeonUpdate()))
    #else
    completion(.success(RegistrationUpdate(state: .unavailable, description: "MWDATCore not linked.")))
    #endif
  }

  func startRegistration(completion: @escaping (Result<Void, Error>) -> Void) {
    #if canImport(MWDATCore)
    do {
      try Wearables.shared.startRegistration()
      completion(.success(()))
    } catch {
      completion(.failure(error))
      flutterApi.onError(errorInfo: ErrorInfo(code: "registration", message: error.localizedDescription))
    }
    #else
    completion(.success(()))
    #endif
  }

  func startUnregistration(completion: @escaping (Result<Void, Error>) -> Void) {
    #if canImport(MWDATCore)
    do {
      try Wearables.shared.startUnregistration()
      completion(.success(()))
    } catch {
      completion(.failure(error))
      flutterApi.onError(errorInfo: ErrorInfo(code: "unregistration", message: error.localizedDescription))
    }
    #else
    completion(.success(()))
    #endif
  }

  func startStream(config: StreamConfig, completion: @escaping (Result<Void, Error>) -> Void) {
    #if canImport(MWDATCore)
    stopStream()

    let deviceSelector = AutoDeviceSelector(wearables: Wearables.shared)
    let streamConfig = StreamSessionConfig(
      videoCodec: .raw,
      resolution: config.quality.toResolution(),
      frameRate: UInt(config.frameRate)
    )
    let session = StreamSession(streamSessionConfig: streamConfig, deviceSelector: deviceSelector)
    streamSession = session

    streamTasks.append(Task { @MainActor [weak self] in
      guard let self else { return }
      for await state in session.statePublisher {
        flutterApi.onStreamStateChanged(state: state.toPigeon())
      }
    })

    streamTasks.append(Task { @MainActor [weak self] in
      guard let self else { return }
      for await frame in session.videoFramePublisher {
        if let uiImage = frame.makeUIImage(),
           let data = uiImage.jpegData(compressionQuality: 0.6) {
          let videoData = VideoFrameData(data: data, width: Int32(uiImage.size.width), height: Int32(uiImage.size.height))
          flutterApi.onVideoFrame(frame: videoData)
        }
      }
    })

    streamTasks.append(Task { @MainActor [weak self] in
      guard let self else { return }
      for await photo in session.photoDataPublisher {
        let photoData = PhotoData(data: photo.data, format: "jpeg")
        flutterApi.onPhotoCaptured(photo: photoData)
      }
    })

    listenerTokens.append(
      session.errorPublisher.listen { [weak self] error in
        self?.flutterApi.onError(errorInfo: ErrorInfo(code: "stream", message: error.description))
      }
    )

    Task {
      await session.start()
    }
    completion(.success(()))
    #else
    completion(.success(()))
    #endif
  }

  func stopStream() {
    #if canImport(MWDATCore)
    streamTasks.forEach { $0.cancel() }
    streamTasks.removeAll()
    listenerTokens.removeAll()
    Task { await streamSession?.stop() }
    streamSession = nil
    flutterApi.onStreamStateChanged(state: .stopped)
    #endif
  }

  func capturePhoto() {
    #if canImport(MWDATCore)
    streamSession?.capturePhoto(format: .jpeg)
    #endif
  }

  #if canImport(MWDATCore)
  private func startRegistrationStream() {
    registrationTask?.cancel()
    registrationTask = Task { @MainActor [weak self] in
      guard let self else { return }
      for await state in Wearables.shared.registrationStateStream() {
        flutterApi.onRegistrationStateChanged(update: state.toPigeonUpdate())
      }
    }
  }
  #endif
}

#if canImport(MWDATCore)
private extension PermissionStatus {
  func toPigeonResult() -> PermissionResult {
    switch self {
    case .granted:
      return PermissionResult(status: .granted, message: nil)
    case .denied:
      return PermissionResult(status: .denied, message: nil)
    case .error(let err):
      return PermissionResult(status: .error, message: err.description)
    @unknown default:
      return PermissionResult(status: .error, message: "Unknown permission status")
    }
  }
}

private extension RegistrationState {
  func toPigeonUpdate() -> RegistrationUpdate {
    switch self {
    case .registered:
      return RegistrationUpdate(state: .registered, description: nil)
    case .registering:
      return RegistrationUpdate(state: .registering, description: nil)
    case .unregistered:
      return RegistrationUpdate(state: .unregistered, description: nil)
    case .unregistering:
      return RegistrationUpdate(state: .unregistering, description: nil)
    case .unavailable:
      return RegistrationUpdate(state: .unavailable, description: nil)
    case .error(let error):
      return RegistrationUpdate(state: .error, description: error.description)
    @unknown default:
      return RegistrationUpdate(state: .error, description: "Unknown registration state")
    }
  }
}

private extension StreamSessionState {
  func toPigeon() -> StreamState {
    switch self {
    case .stopped:
      return .stopped
    case .waitingForDevice:
      return .waitingForDevice
    case .starting:
      return .starting
    case .streaming:
      return .streaming
    case .stopping:
      return .stopping
    case .paused:
      return .paused
    @unknown default:
      return .stopped
    }
  }
}

private extension VideoQuality {
  func toResolution() -> StreamingResolution {
    switch self {
    case .low:
      return .low
    case .medium:
      return .medium
    case .high:
      return .high
    }
  }
}
#endif

File: android/src/main/kotlin/com/chunkytofustudios/meta_wearables/MetaWearablesPlugin.kt
package com.chunkytofustudios.meta_wearables

import android.app.Activity
import android.content.Context
import androidx.activity.result.ActivityResultLauncher
import com.meta.wearable.dat.camera.StreamSession
import com.meta.wearable.dat.camera.types.PhotoData
import com.meta.wearable.dat.camera.types.StreamConfiguration
import com.meta.wearable.dat.camera.types.StreamSessionState
import com.meta.wearable.dat.camera.types.VideoFrame
import com.meta.wearable.dat.camera.types.VideoQuality
import com.meta.wearable.dat.core.Wearables
import com.meta.wearable.dat.core.selectors.AutoDeviceSelector
import com.meta.wearable.dat.core.types.Permission as SdkPermission
import com.meta.wearable.dat.core.types.PermissionStatus as SdkPermissionStatus
import com.meta.wearable.dat.core.types.RegistrationState
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancel
import kotlinx.coroutines.flow.collectLatest
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.launch
import kotlinx.coroutines.suspendCancellableCoroutine
import java.nio.ByteBuffer
import kotlin.coroutines.resume

/** MetaWearablesPlugin */
class MetaWearablesPlugin : FlutterPlugin, ActivityAware {
    private var activity: Activity? = null
    private var context: Context? = null
    private var hostApi: MetaWearablesHostApiImpl? = null

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        context = binding.applicationContext
        val flutterApi = WearablesFlutterApi(binding.binaryMessenger)
        hostApi = MetaWearablesHostApiImpl(binding.applicationContext, flutterApi)
        WearablesHostApi.setUp(binding.binaryMessenger, hostApi)
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        WearablesHostApi.setUp(binding.binaryMessenger, null)
        hostApi?.dispose()
        hostApi = null
        context = null
    }

    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        activity = binding.activity
        hostApi?.setActivity(binding.activity, binding)
    }

    override fun onDetachedFromActivity() {
        activity = null
        hostApi?.clearActivity()
    }

    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
        onAttachedToActivity(binding)
    }

    override fun onDetachedFromActivityForConfigChanges() {
        onDetachedFromActivity()
    }
}

private class MetaWearablesHostApiImpl(
    private val appContext: Context,
    private val flutterApi: WearablesFlutterApi,
) : WearablesHostApi {
    private val scope: CoroutineScope = CoroutineScope(SupervisorJob() + Dispatchers.Main)
    private var registrationJob: Job? = null
    private var streamStateJob: Job? = null
    private var videoJob: Job? = null
    private var currentStream: StreamSession? = null
    private var permissionLauncher: ActivityResultLauncher<Permission>? = null
    private var permissionContinuation: (SdkPermissionStatus) -> Unit = {}

    fun setActivity(activity: Activity, binding: ActivityPluginBinding) {
        // DAT exposes a RequestPermissionContract; wrap via ActivityResultContracts to keep control
        permissionLauncher =
            activity.registerForActivityResult(Wearables.RequestPermissionContract()) { status ->
                permissionContinuation(status)
            }
    }

    fun clearActivity() {
        permissionLauncher = null
        permissionContinuation = {}
    }

    fun dispose() {
        scope.cancel()
        currentStream?.close()
    }

    override fun initialize() {
        Wearables.initialize(appContext)
        registrationJob?.cancel()
        registrationJob =
            scope.launch {
                Wearables.registrationState.collectLatest { state ->
                    flutterApi.onRegistrationStateChanged(
                        state.toPigeonUpdate()
                    )
                }
            }
    }

    override fun checkPermission(
        permission: com.chunkytofustudios.meta_wearables.meta_wearables.Permission,
        callback: (Result<PermissionResult>) -> Unit
    ) {
        scope.launch {
            try {
                val status = Wearables.checkPermissionStatus(permission.toSdk())
                callback(Result.success(status.toPigeonResult()))
            } catch (t: Throwable) {
                callback(Result.failure(t))
                flutterApi.onError(ErrorInfo("permission", t.message ?: "Unknown error"))
            }
        }
    }

    override fun requestPermission(
        permission: com.chunkytofustudios.meta_wearables.meta_wearables.Permission,
        callback: (Result<PermissionResult>) -> Unit
    ) {
        val launcher = permissionLauncher
        if (launcher == null) {
            callback(
                Result.failure(
                    FlutterError("no-activity", "Activity not attached; cannot request permission.", null)
                )
            )
            return
        }
        scope.launch {
            val status =
                suspendCancellableCoroutine { cont ->
                    permissionContinuation = { result ->
                        if (cont.isActive) cont.resume(result)
                    }
                    launcher.launch(permission.toSdk())
                    cont.invokeOnCancellation { permissionContinuation = {} }
                }
            callback(Result.success(status.toPigeonResult()))
        }
    }

    override fun getRegistrationState(callback: (Result<RegistrationUpdate>) -> Unit) {
        scope.launch {
            try {
                val state = Wearables.registrationState.first()
                callback(Result.success(state.toPigeonUpdate()))
            } catch (t: Throwable) {
                callback(Result.failure(t))
            }
        }
    }

    override fun startRegistration(callback: (Result<Unit>) -> Unit) {
        scope.launch {
            try {
                Wearables.startRegistration(appContext)
                callback(Result.success(Unit))
            } catch (t: Throwable) {
                callback(Result.failure(t))
                flutterApi.onError(ErrorInfo("registration", t.message ?: "Unknown error"))
            }
        }
    }

    override fun startUnregistration(callback: (Result<Unit>) -> Unit) {
        scope.launch {
            try {
                Wearables.startUnregistration(appContext)
                callback(Result.success(Unit))
            } catch (t: Throwable) {
                callback(Result.failure(t))
                flutterApi.onError(ErrorInfo("unregistration", t.message ?: "Unknown error"))
            }
        }
    }

    override fun startStream(config: StreamConfig, callback: (Result<Unit>) -> Unit) {
        scope.launch {
            try {
                stopStreamInternal()
                val streamConfig =
                    StreamConfiguration(
                        videoQuality = config.quality.toSdk(),
                        config.frameRate
                    )
                currentStream =
                    Wearables.startStreamSession(
                        appContext,
                        AutoDeviceSelector(),
                        streamConfig
                    )
                currentStream?.let { session ->
                    streamStateJob =
                        launch {
                            session.state.collectLatest { state ->
                                flutterApi.onStreamStateChanged(state.toPigeon())
                                if (state == StreamSessionState.STOPPED) {
                                    stopStreamInternal()
                                }
                            }
                        }
                    videoJob =
                        launch {
                            session.videoStream.collectLatest { frame ->
                                flutterApi.onVideoFrame(frame.toPigeon())
                            }
                        }
                }
                callback(Result.success(Unit))
            } catch (t: Throwable) {
                flutterApi.onError(ErrorInfo("startStream", t.message ?: "Unknown error"))
                callback(Result.failure(t))
            }
        }
    }

    override fun stopStream() {
        stopStreamInternal()
    }

    override fun capturePhoto() {
        scope.launch {
            try {
                currentStream?.capturePhoto()?.onSuccess { photo ->
                    flutterApi.onPhotoCaptured(photo.toPigeon())
                }?.onFailure {
                    flutterApi.onError(ErrorInfo("capturePhoto", it.message ?: "Unknown error"))
                }
            } catch (t: Throwable) {
                flutterApi.onError(ErrorInfo("capturePhoto", t.message ?: "Unknown error"))
            }
        }
    }

    private fun stopStreamInternal() {
        videoJob?.cancel()
        streamStateJob?.cancel()
        currentStream?.close()
        currentStream = null
        flutterApi.onStreamStateChanged(StreamState.STOPPED)
    }
}

private fun SdkPermissionStatus.toPigeonResult(): PermissionResult =
    when (this) {
        SdkPermissionStatus.Granted -> PermissionResult(status = PermissionStatus.GRANTED)
        SdkPermissionStatus.Denied -> PermissionResult(status = PermissionStatus.DENIED)
        is SdkPermissionStatus.Error -> PermissionResult(
            status = PermissionStatus.ERROR,
            message = this.error.description
        )
    }

private fun com.chunkytofustudios.meta_wearables.meta_wearables.Permission.toSdk(): SdkPermission =
    when (this) {
        com.chunkytofustudios.meta_wearables.meta_wearables.Permission.CAMERA -> SdkPermission.CAMERA
    }

private fun RegistrationState.toPigeonUpdate(): RegistrationUpdate =
    when (this) {
        is RegistrationState.Available -> RegistrationUpdate(
            state = RegistrationStatePigeon.AVAILABLE,
            errorCode = error?.name,
            description = error?.description
        )
        is RegistrationState.Unavailable -> RegistrationUpdate(
            state = RegistrationStatePigeon.UNAVAILABLE,
            errorCode = error?.name,
            description = error?.description
        )
        is RegistrationState.Registering -> RegistrationUpdate(
            state = RegistrationStatePigeon.REGISTERING,
            errorCode = error?.name,
            description = error?.description
        )
        is RegistrationState.Registered -> RegistrationUpdate(
            state = RegistrationStatePigeon.REGISTERED,
            errorCode = error?.name,
            description = error?.description
        )
        is RegistrationState.Unregistering -> RegistrationUpdate(
            state = RegistrationStatePigeon.UNREGISTERING,
            errorCode = error?.name,
            description = error?.description
        )
    }

private fun StreamSessionState.toPigeon(): StreamState =
    when (this) {
        StreamSessionState.STARTING -> StreamState.STARTING
        StreamSessionState.STARTED -> StreamState.STARTED
        StreamSessionState.STREAMING -> StreamState.STREAMING
        StreamSessionState.STOPPING -> StreamState.STOPPING
        StreamSessionState.STOPPED -> StreamState.STOPPED
        StreamSessionState.CLOSED -> StreamState.CLOSED
    }

private fun VideoQualityPigeon.toSdk(): VideoQuality =
    when (this) {
        VideoQualityPigeon.LOW -> VideoQuality.LOW
        VideoQualityPigeon.MEDIUM -> VideoQuality.MEDIUM
        VideoQualityPigeon.HIGH -> VideoQuality.HIGH
    }

private fun VideoFrame.toPigeon(): VideoFrameData {
    val buffer: ByteBuffer = this.buffer
    val data = ByteArray(buffer.remaining())
    val position = buffer.position()
    buffer.get(data)
    buffer.position(position)
    return VideoFrameData(data = data, width = width, height = height)
}

private fun PhotoData.toPigeon(): PhotoDataPigeon =
    when (this) {
        is PhotoData.Bitmap -> {
            val stream = java.io.ByteArrayOutputStream()
            bitmap.compress(android.graphics.Bitmap.CompressFormat.JPEG, 90, stream)
            PhotoDataPigeon(data = stream.toByteArray(), format = "jpeg")
        }
        is PhotoData.HEIC -> {
            val byteArray = ByteArray(data.remaining())
            data.get(byteArray)
            PhotoDataPigeon(data = byteArray, format = "heic")
        }
    }

// Pigeon generated enums are in the same package; use aliases for clarity
private typealias RegistrationStatePigeon = com.chunkytofustudios.meta_wearables.meta_wearables.RegistrationState
private typealias VideoQualityPigeon = com.chunkytofustudios.meta_wearables.meta_wearables.VideoQuality
